endpoints/put_test.go

```diff
@@ -0,0 +1,1093 @@
+		},
+	}
+	for _, tc := range testCases {
+		// set test
+		putHandler := &PutHandler{
+			memory: syncPools{
+				requestPool: sync.Pool{
+					New: func() interface{} { return &putRequest{} },
+				},
+			},
+			cfg: putHandlerConfig{maxNumValues: 1},
+		}
+		// run
+		put, err := putHandler.parseRequest(tc.getInputRequest())
+		// assertions
+		assert.Equal(t, tc.expected.put, put, tc.desc)
+		assert.Equal(t, tc.expected.err, err, tc.desc)
+	}
+}
+
+// TestParsePutObject asserts *PutHandler's parsePutObject(p PutObject) method
+func TestParsePutObject(t *testing.T) {
+	type testOut struct {
+		value string
+		err   error
+	}
+	testCases := []struct {
+		desc     string
+		in       putObject
+		expected testOut
+	}{
+		{
+			"empty value, expect error",
+			putObject{},
+			testOut{
+				value: "",
+				err:   utils.NewPBCError(utils.MISSING_VALUE),
+			},
+		},
+		{
+			"negative time-to-live, expect error",
+			putObject{
+				TTLSeconds: -1,
+				Value:      json.RawMessage(`<tag>Your XML content goes here.</tag>`),
+			},
+			testOut{
+				value: "",
+				err:   utils.NewPBCError(utils.NEGATIVE_TTL, "ttlseconds must not be negative -1."),
+			},
+		},
+		{
+			"non xml nor json type, expect error",
+			putObject{
+				Type:       "unknown",
+				TTLSeconds: 60,
+				Value:      json.RawMessage(`<tag>Your XML content goes here.</tag>`),
+			},
+			testOut{
+				value: "",
+				err:   utils.NewPBCError(utils.UNSUPPORTED_DATA_TO_STORE, "Type must be one of [\"json\", \"xml\"]. Found unknown"),
+			},
+		},
+		{
+			"xml type value is not a string, expect error",
+			putObject{
+				Type:       "xml",
+				TTLSeconds: 60,
+				Value:      json.RawMessage(`<tag>XML</tag>`),
+			},
+			testOut{
+				value: "",
+				err:   utils.NewPBCError(utils.MALFORMED_XML, "XML messages must have a String value. Found [60 116 97 103 62 88 77 76 60 47 116 97 103 62]"),
+			},
+		},
+		{
+			"xml type value is surrounded by quotes and, therefore, a string. No errors expected",
+			putObject{
+				Type:       "xml",
+				TTLSeconds: 60,
+				Value:      json.RawMessage(`"<tag>XML</tag>"`),
+			},
+			testOut{
+				"xml<tag>XML</tag>",
+				nil,
+			},
+		},
+		{
+			"valid JSON input, no errors expected",
+			putObject{
+				Type:       "json",
+				TTLSeconds: 60,
+				Value:      json.RawMessage(`{"native":"{\"context\":1,\"plcmttype\":1,\"assets\":[{\"img\":{\"wmin\":30}}]}}`),
+			},
+			testOut{
+				`json{"native":"{\"context\":1,\"plcmttype\":1,\"assets\":[{\"img\":{\"wmin\":30}}]}}`,
+				nil,
+			},
+		},
+	}
+	for _, tc := range testCases {
+		// run
+		actualPutString, actualError := parsePutObject(tc.in)
+
+		// assertions
+		assert.Equal(t, tc.expected.value, actualPutString, tc.desc)
+		assert.Equal(t, tc.expected.err, actualError, tc.desc)
+	}
+}
+
+// TestLogBackendError asserts this package's logBackendError(err error, index int) function
+func TestLogBackendError(t *testing.T) {
+	type testOutput struct {
+		err  error
+		code int
```

Comment:
The struct field `code` is not used so it can be deleted.

Reply:
I forgot to assert the error code. Do you mind if we add this assertion to the test?